functions/executor/call.py

```diff
@@ -104,33 +111,183 @@ def exec_dag_function(pusher_cache, kvs, triggers, function, schedule):
+
+            for key in dependency:
+                dep = new_trigger.dependency.add()
+                dep.key = key
+                dep.vector_clock = dependency[key]
+
+            dest_ip = schedule.locations[conn.sink]
+            sckt = pusher_cache.get(sutils._get_dag_trigger_address(dest_ip))
+            sckt.send(new_trigger.SerializeToString())
+
+    if is_sink:
+        logging.info('DAG %s (ID %d) completed in causal mode; result at %s.' %
+                (schedule.dag.name, trigger.id, schedule.response_id))
+
+        vector_clock = {}
+        if schedule.response_id in dependency:
+            if schedule.id in dependency[schedule.response_id]:
+                dependency[schedule.response_id][schedule.id] += 1
+            else:
+                dependency[schedule.response_id][schedule.id] = 1
+            vector_clock = dependency[schedule.response_id][schedule.id]
+            del dependency[schedule.response_id]
+        else:
+            vector_clock = {schedule.id : 1}
+
+        succeed = kvs.causal_put(schedule.response_id, vector_clock, dependency, serialize_val(result), schedule.id)
+        while not succeed:
+            kvs.causal_put(schedule.response_id, vector_clock, dependency, serialize_val(result), schedule.id)
+
+
+
+def _exec_func_causal(kvs, func, args, kv_pairs, schedule, address_to_versioned_key_list_map):
     func_args = ()
 
+    to_resolve = []
+
     # resolve any references to KVS objects
-    for arg in args:
+    key_index_map = {}
+    for i, arg in enumerate(args):
         if isinstance(arg, FluentReference):
-            func_args += (_resolve_ref(arg, kvs),)
-        else:
-            func_args += (arg,)
+            to_resolve.append(arg)
+            key_index_map[arg.key] = i
+        func_args += (arg,)
+
+    if len(to_resolve) > 0: 
+        _resolve_ref_causal(to_resolve, kvs, kv_pairs, schedule, address_to_versioned_key_list_map)
+
+        for key in kv_pairs:
+            func_args[key_index_map[key]] = kv_pairs[key][1]
 
     # execute the function
     return  func(*func_args)
 
-def _resolve_ref(ref, kvs):
-    ref_data = kvs.get(ref.key, ref.obj_type)
 
-    # when chaining function executions, we must wait
-    while not ref_data:
-        ref_data = kvs.get(ref.key, ref.obj_type)
+def _resolve_ref_causal(refs, kvs, kv_pairs, schedule, address_to_versioned_key_list_map):
+    future_read_set = _compute_children_read_set(schedule)
+    result = kvs.causal_get(refs, future_read_set, schedule.consistency, schedule.id)
+
+    while not result:
+        result = kvs.causal_get(refs, future_read_set, address_to_versioned_key_list_map, schedule.consistency, schedule.id)
 
-    if ref.deserialize:
-        ref_data = deserialize_val(ref_data.reveal()[1])
+    if result[0] is not None:
+        address_to_versioned_key_list_map[result[0][0]] = list(result[0][1])
 
-    return ref_data
+    kv_pairs = result[1]
```

Comment:
Also not deserializing values here. Will break.

Reply:
fixed